### PR TITLE
Re-jig default region logic a bit

### PR DESCRIPTION
**Pull request by kiiadi**
Removing deprecated 'getById' region method, and changing default logic to work regardless of partition

Related to #1756

### CODE REVIEW

File: core/src/software/aws/toolkits/core/utils/NullabilityUtils.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.core.utils
+
+/**
+ * Invokes the [block] if the receiving type is null
+ */
+fun <T : Any?> T.onNull(block: () -> Unit): T {
```
Comment (abrooksv) [inline core/src/software/aws/toolkits/core/utils/NullabilityUtils.kt:9]: The opposite of `let`.... hmmm

I like.

`?:` doesn't let you do a code block

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/region/AwsRegionProvider.kt [hunk 2]
```diff
@@ -14,10 +15,16 @@ import software.aws.toolkits.core.region.ServiceEndpointResource
...
 import software.aws.toolkits.core.utils.getLogger
 import software.aws.toolkits.core.utils.inputStream
+import software.aws.toolkits.core.utils.onNull
+import software.aws.toolkits.core.utils.tryOrNull
 import software.aws.toolkits.core.utils.warn
 import software.aws.toolkits.jetbrains.core.RemoteResourceResolverProvider
 
 class AwsRegionProvider constructor(remoteResourceResolverProvider: RemoteResourceResolverProvider) : ToolkitRegionProvider() {
+    private val regionChain by lazy {
+        // Querying the instance metadata is expensive due to high timeouts and retries
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/region/AwsRegionProvider.kt:25]: They fixed this in the SDK I believe, we should recheck this one day...

File: core/src/software/aws/toolkits/core/utils/NullabilityUtils.kt (hunk 1 above)
Comment (kiiadi) [reply core/src/software/aws/toolkits/core/utils/NullabilityUtils.kt:9]: `?:` still evaluates this to determine the overall statement result so you get `T|Unit` or something weird - I wanted to maintain the fact it's `null` but do something in that case.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/region/AwsRegionProvider.kt (hunk 2 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/region/AwsRegionProvider.kt:25]: Yeah and if it isn't fixed - do it manually - I want us to create an EC2 credential provider too - cos I think EC2 should "just work"

File: core/src/software/aws/toolkits/core/utils/NullabilityUtils.kt (hunk 1 above)
Comment (abrooksv) [reply core/src/software/aws/toolkits/core/utils/NullabilityUtils.kt:9]: Ya, I like the addition.

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/region/AwsRegionProvider.kt [hunk 3]
```diff
@@ -33,25 +40,24 @@ class AwsRegionProvider constructor(remoteResourceResolverProvider: RemoteResour
...
 
-    private fun fallbackRegion(): AwsRegion = regions(DEFAULT_PARTITION).getOrElse(DEFAULT_REGION) {
-        throw IllegalStateException("Region provider data is missing default data")
+    override fun defaultRegion(): AwsRegion {
+        val regionIdFromChain = LOG.tryOrNull("Failed to find default region in chain", level = Level.WARN) {
+            regionChain.region.id()
+        }
+
+        val regionFromChain = regionIdFromChain?.let { regionId ->
+            this[regionId].onNull { LOG.warn { "Could not find $regionId in endpoint data" } }
```
Comment (abrooksv) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/region/AwsRegionProvider.kt:53]: We do have `Logger.logWhenNull`

Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/region/AwsRegionProvider.kt:53]: But maybe this way reads better actually... No strong feelings

Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/region/AwsRegionProvider.kt:53]: oooh...

**Comment by kiiadi** [comment]
Collapsed this down into https://github.com/aws/aws-toolkit-jetbrains/pull/1756

**Pull request closed without merging**